NthKey/Model/SubscriptionManager.swift

```diff
@@ -0,0 +1,246 @@
+
+                if sharedSecret != nil {
+                    validateReceipt()
+                } else {
+                    checkExpirationDateFromPayment(transaction.payment)
+                }
+            case .restored:
+                totalRestoredPurchases += 1
+                purchasePublisher.send(("Restored ",true))
+
+                if sharedSecret != nil {
+                    validateReceipt()
+                } else {
+                    checkExpirationDateFromPayment(transaction.payment)
+                }
+            case .failed:
+                if let error = transaction.error as? SKError {
+                    purchasePublisher.send(("Payment Error \(error.code) ",false))
+                    print("Payment Failed \(error.code)")
+                }
+            case .deferred:
+                print("Ask Mom ...")
+                purchasePublisher.send(("Payment Diferred ",false))
+            case .purchasing:
+                print("working on it...")
+                purchasePublisher.send(("Payment in Process ",false))
+            default:
+                break
+            }
+
+            guard !(transaction.transactionState == .purchasing || transaction.transactionState == .deferred) else {
+                return
+            }
+            
+            queue.finishTransaction(transaction)
+        }
+    }
+
+    // Sent when all transactions from the user's purchase history have successfully been added back to the queue.
+    func paymentQueueRestoreCompletedTransactionsFinished(_ queue: SKPaymentQueue) {
+        guard totalRestoredPurchases != 0 else {
+            purchasePublisher.send(("IAP: No purchases to restore!",true))
+            return
+        }
+
+        purchasePublisher.send(("IAP: Purchases successfull restored!",true))
+    }
+
+    // Sent when an error is encountered while adding transactions from the user's purchase history back to the queue.
+    func paymentQueue(_ queue: SKPaymentQueue, restoreCompletedTransactionsFailedWithError error: Error) {
+        guard let error = error as? SKError else { return }
+        let reason = error.code != .paymentCancelled ? " Restore" : ""
+        purchasePublisher.send(("IAP\(reason) Error: " + error.localizedDescription, false))
+    }
+
+    // TODO: Possible should implement for any transactions which will be revoked
+    // func paymentQueue(_ queue: SKPaymentQueue, removedTransactions transactions: [SKPaymentTransaction])
+}
+
+extension SubscriptionManager: SKRequestDelegate {
+    func requestDidFinish(_ request: SKRequest) {
+        print("request did finish")
+    }
+
+    func request(_ request: SKRequest, didFailWithError error: Error) {
+        purchasePublisher.send(("Purchase request failed ",true))
+        print("didFailWithError ",error)
+    }
+}
+
+#if DEBUG
+extension SubscriptionManager {
+    static var mock: SubscriptionManager = SubscriptionManager(identifiers: ["com.nthkey.monthly", "com.nthkey.annual"])
```

Comment:
Forgot to drop annual here?

Reply:
No, we can ask AppStore for any iAP products (it used for debug and preview). If some of product isn't in the store now then we'll receive info about it in debug https://github.com/Sjors/nthkey-ios/blob/8feb68d44d7544e8506ce74dfc68756ebcb703e2/NthKey/Model/SubscriptionManager.swift#L103